kairos-data/src/transaction.rs

```diff
@@ -0,0 +1,143 @@
+use crate::schema::transactions;
+use bigdecimal::BigDecimal;
+use chrono::{NaiveDateTime, Utc};
+use diesel::prelude::*;
+use hex;
+use serde::{Deserialize, Serialize};
+
+use kairos_circuit_logic::transactions::{
+    KairosTransaction, L1Deposit, Signed, Transfer, Withdraw,
+};
+
+#[derive(diesel_derive_enum::DbEnum, Debug, Serialize)]
+#[ExistingTypePath = "crate::schema::sql_types::Transaction"]
+pub enum Transaction {
+    Deposit,
+    Transfer,
+    Withdrawal,
+}
+
+#[derive(Queryable, Debug, Identifiable, Insertable, Serialize, Selectable)]
+#[diesel(primary_key(timestamp, amount, public_key))]
+#[diesel(table_name = transactions)]
+#[diesel(check_for_backend(diesel::pg::Pg))]
+pub struct Transactions {
+    pub timestamp: NaiveDateTime,
+    pub public_key: String,
+    pub nonce: Option<BigDecimal>,
+    pub trx: Transaction,
+    pub amount: BigDecimal,
+    pub recipient: Option<String>,
+}
+
+#[derive(Deserialize)]
+pub struct TransactionFilter {
+    pub sender: Option<String>,
+    pub min_timestamp: Option<NaiveDateTime>,
+    pub max_timestamp: Option<NaiveDateTime>,
+    pub min_amount: Option<i64>,
+    pub max_amount: Option<i64>,
+    pub recipient: Option<String>,
+}
+
+pub async fn get(
+    pool: &crate::Pool,
+    filter: TransactionFilter,
+) -> Result<Vec<Transactions>, crate::errors::DBError> {
+    let conn = pool.get().await?;
+    let res = conn
+        .interact(move |conn| {
+            let mut query = transactions::table.into_boxed::<diesel::pg::Pg>();
+
+            if let Some(sender) = filter.sender {
+                query = query.filter(transactions::public_key.eq(sender));
+            }
+            if let Some(min_timestamp) = filter.min_timestamp {
+                query = query.filter(transactions::timestamp.ge(min_timestamp));
+            }
+            if let Some(max_timestamp) = filter.max_timestamp {
+                query = query.filter(transactions::timestamp.le(max_timestamp));
+            }
+            if let Some(min_amount) = filter.min_amount {
+                query = query.filter(transactions::amount.ge(BigDecimal::from(min_amount)));
+            }
+            if let Some(max_amount) = filter.max_amount {
+                query = query.filter(transactions::amount.le(BigDecimal::from(max_amount)));
+            }
+            if let Some(recipient) = filter.recipient {
+                query = query.filter(transactions::recipient.eq(recipient));
+            }
+
+            query
+                .select(Transactions::as_select())
+                .limit(500)
+                .load::<Transactions>(conn)
+        })
+        .await??;
+    Ok(res)
+}
+
+pub async fn insert(
+    pool: crate::Pool,
```

Comment:
Use a reference here too.

Reply:
In general, I think the right approach is to take references for complex parameter datatypes until ownership is strictly necessary. That way we can avoid clones.